wallets/dcr/sync.go

```diff
@@ -0,0 +1,484 @@
+	for _, listener := range wallet.syncData.SyncProgressListeners {
+		listeners = append(listeners, listener)
+	}
+
+	return listeners
+}
+
+func (wallet *Wallet) PublishLastSyncProgress(uniqueIdentifier string) error {
+	wallet.syncData.mu.RLock()
+	defer wallet.syncData.mu.RUnlock()
+
+	syncProgressListener, exists := wallet.syncData.SyncProgressListeners[uniqueIdentifier]
+	if !exists {
+		return errors.New(ErrInvalid)
+	}
+
+	if wallet.syncData.syncing && wallet.syncData.activeSyncData != nil {
+		switch wallet.syncData.activeSyncData.syncStage {
+		case HeadersFetchSyncStage:
+			syncProgressListener.OnHeadersFetchProgress(&wallet.syncData.headersFetchProgress)
+		case AddressDiscoverySyncStage:
+			syncProgressListener.OnAddressDiscoveryProgress(&wallet.syncData.addressDiscoveryProgress)
+		case HeadersRescanSyncStage:
+			syncProgressListener.OnHeadersRescanProgress(&wallet.syncData.headersRescanProgress)
+		}
+	}
+
+	return nil
+}
+
+func (wallet *Wallet) EnableSyncLogs() {
+	wallet.syncData.mu.Lock()
+	wallet.syncData.showLogs = true
+	wallet.syncData.mu.Unlock()
+}
+
+func (wallet *Wallet) SyncInactiveForPeriod(totalInactiveSeconds int64) {
+	wallet.syncData.mu.Lock()
+	defer wallet.syncData.mu.Unlock()
+
+	if !wallet.syncData.syncing || wallet.syncData.activeSyncData == nil {
+		log.Debug("Not accounting for inactive time, wallet is not syncing.")
+		return
+	}
+
+	wallet.syncData.totalInactiveSeconds += totalInactiveSeconds
+	if wallet.syncData.connectedPeers == 0 {
+		// assume it would take another 60 seconds to reconnect to peers
+		wallet.syncData.totalInactiveSeconds += 60
+	}
+}
+
+func (wallet *Wallet) SpvSync() error {
+	// prevent an attempt to sync when the previous syncing has not been canceled
+	if wallet.IsSyncing() || wallet.IsSynced() {
+		return errors.New(ErrSyncAlreadyInProgress)
+	}
+
+	addr := &net.TCPAddr{IP: net.ParseIP("::1"), Port: 0}
+	addrManager := addrmgr.New(wallet.rootDir, net.LookupIP) // TODO: be mindful of tor
+	lp := p2p.NewLocalPeer(wallet.chainParams, addr, addrManager)
+
+	var validPeerAddresses []string
+	peerAddresses := wallet.ReadStringConfigValueForKey(SpvPersistentPeerAddressesConfigKey, "")
+	if peerAddresses != "" {
+		addresses := strings.Split(peerAddresses, ";")
+		for _, address := range addresses {
+			peerAddress, err := NormalizeAddress(address, wallet.chainParams.DefaultPort)
+			if err != nil {
+				log.Errorf("SPV peer address(%s) is invalid: %v", peerAddress, err)
+			} else {
+				validPeerAddresses = append(validPeerAddresses, peerAddress)
+			}
+		}
+
+		if len(validPeerAddresses) == 0 {
+			return errors.New(ErrInvalidPeers)
+		}
+	}
+
+	// init activeSyncData to be used to hold data used
+	// to calculate sync estimates only during sync
+	wallet.initActiveSyncData()
+
+	wallets := make(map[int]*w.Wallet)
+	wallets[0] = wallet.Internal()
+	wallet.WaitingForHeaders = true
+	wallet.Syncing = true
+
+	syncer := spv.NewSyncer(wallets, lp)
```

Comment:
This looks better to me:
`syncer := spv.NewSyncer(map[int]*w.Wallet{0:wallet.Internal()}, lp)`

Reply:
as of now, i'm not writing any new code, i just moved all the old code into a package, perhaps we can consider this during a code cleanup